### ISSUE TITLE
Callback race condition for non-resumable file uploads

### ISSUE THREAD

**Issue by rraina** [googleapis/nodejs-storage#408]
#### Environment details

  - OS: `MacOS X 10.13.4`
  - Node.js version: `8.12.0`
  - npm version: `6.4.1`
  - `@google-cloud/storage` version: 1.7.0 and 2.0.3

#### Steps to reproduce

  1. Create a dummy file to upload (smaller is better)
```
const fs = require('fs-extra');

const localFileName = 'sample.txt';
const data = '0'.repeat(1000);

await fs.writeFile(localFileName, data, { encoding: 'utf8' } );
``` 
  2. Upload file to GStorage
```
const bucket = await GStorage.createBucket('node-sample-bucket');
bucket.upload(localFileName, {
  destination: destFileName,
  metadata: { contentType: 'text/plain'},
});
```

3. Delete local file
```
await fs.unlink(localFileName);
```

#### NOTES:
- The main issue is that there is a race condition between the fs.stat callback https://github.com/googleapis/nodejs-storage/blob/197b7602d69bb3152ae363a3bbd2b55f83b5a96f/src/bucket.ts#L2630 and the pipe callback when stream has ended https://github.com/googleapis/nodejs-storage/blob/197b7602d69bb3152ae363a3bbd2b55f83b5a96f/src/bucket.ts#L2646
- As a user, I expect no further system calls to my local file once the upload Promise/callback has returned successfully. The only exception is when I have set `resumable` to true.
- The temporary workaround is to manually set `resumable` to `false`; however, I don't think this is a good user experience, and can lead to a lot of developer frustration.
- Happy to make a PR to fix issue if it will be merged in!

**Comment by stephenplusplus** [comment]
The code seems correct. Note the 1) and 2).

```js
if (is.boolean(options.resumable)) {
  upload();
} else {
  // Determine if the upload should be resumable if it's over the threshold.
  fs.stat(pathString, (err, fd) => {
    if (err) {
      callback(err);
      return;
    }

    options.resumable = fd.size > RESUMABLE_THRESHOLD;

    // ------
    // 1) Upload is called
    // ------
    upload();
  });
}

function upload() {
  // ------
  // 2) Upload begins, when `callback` is used again
  // ------
  fs.createReadStream(pathString)
      .on('error', callback)
      .pipe(newFile.createWriteStream(options))
      .on('error', callback)
      .on('finish', () => {
        callback(null, newFile, newFile.metadata);
      });
}
```

I'm going to close the issue, but just let me know if you've experienced an error that suggests there's a bug somewhere.

Thanks!

**Comment by rraina** [comment]
Hey Stephen,

Thanks for reviewing. It does seem like the callback ordering is correct.

With that said, on version 1.7.0, I still had a race condition where the open syscall was being called despite me having already finished the upload.

Are you able to replicate the issue using steps I mentioned in the above ticket?

Thanks,
Rahul

On Sep 29, 2018, at 6:57 AM, Stephen <notifications@github.com<mailto:notifications@github.com>> wrote:


The code seems correct. Note the 1) and 2).

if (is.boolean(options.resumable)) {
  upload();
} else {
  // Determine if the upload should be resumable if it's over the threshold.
  fs.stat(pathString, (err, fd) => {
    if (err) {
      callback(err);
      return;
    }

    options.resumable = fd.size > RESUMABLE_THRESHOLD;

    // ------
    // 1) Upload is called
    // ------
    upload();
  });
}

function upload() {
  // ------
  // 2) Upload begins, when `callback` is used again
  // ------
  fs.createReadStream(pathString)
      .on('error', callback)
      .pipe(newFile.createWriteStream(options))
      .on('error', callback)
      .on('finish', () => {
        callback(null, newFile, newFile.metadata);
      });
}

I'm going to close the issue, but just let me know if you've experienced an error that suggests there's a bug somewhere.

Thanks!

**Comment by stephenplusplus** [comment]
Hmm, maybe you need to use an `await` in front of `bucket.upload(...`? That method expects a callback, otherwise, it returns a promise.

**Comment by rraina** [comment]
I did! Were you not able to replicate the issue?

Thanks,
Rahul

On Sep 30, 2018, at 9:54 AM, Stephen <notifications@github.com<mailto:notifications@github.com>> wrote:


Hmm, maybe you need to use an await in front of bucket.upload(...? That method expects a callback, otherwise, it returns a promise.

**Comment by stephenplusplus** [comment]
When there is no "await" in front of "bucket.upload()", I get:

```
Error: ENOENT: no such file or directory, open '{{the filepath}}'
```

When I put the `await` in, I don't receive an error.

I asked about await, because it is missing from the code you shared in your reproduction steps.

**Comment by rraina** [comment]
Found the issue. You were correct.

I had wrapped this call in a function, and forgot to return the promise. Therefore, even though I was awaiting the wrapped call, it was breaking.

Thanks, and sorry for the goose chase!

On Sep 30, 2018, at 10:12 AM, Stephen <notifications@github.com<mailto:notifications@github.com>> wrote:


When there is no "await" in front of "bucket.upload()", I get:

Error: ENOENT: no such file or directory, open '{{the filepath}}'


When I put the await in, I don't receive an error.

I asked about await, because it is missing from the code you shared in your reproduction steps.
